### PR TITLE
redo module to cater to managing tfstate in remote accounts

### PR DESCRIPTION
**Pull request by cdaniluk**
This is the way tf recommends now, and it seems like a perfectly fine way to go to me compared to the dumpster fire that is the way we used to do it

### CODE REVIEW

File: assumerole.tf [hunk 1]
```diff
@@ -0,0 +1,36 @@
+locals {
+  external_id = coalesce(var.assumerole_role_external_id, random_password.external_id.result)
+}
+
+resource "local_file" "assumerole_addrole" {
+  count = var.create_assumerole_template ? 1 : 0
+
+  filename = "assumerole/addrole.sh"
```
Comment (sblack4) [inline assumerole.tf:8]: when does this get run?

File: assumerole.tf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  })
+
+}
+
+resource "local_sensitive_file" "assumerole_tfassumerole" {
+  count = var.create_assumerole_template ? 1 : 0
+
+  filename = "assumerole/tfassumerole.cfn.yml"
+
+  content = templatefile("${path.module}/template/tfassumerole.cfn.yml.tftpl", {
```
Comment (sblack4) [inline assumerole.tf:21]: why use CFN for this? why not more TF?

Comment (sblack4) [reply assumerole.tf:21]: or why not make the whole thing CFN to avoid the chicken-and-egg problem?

File: variables.tf [hunk 3]
```diff
@@ -1,52 +1,95 @@
...
   type        = string
 }
 
+variable "lifecycle_rules" {
+  description = "lifecycle rules to apply to the bucket (set to null to skip lifecycle rules)"
+
+  default = [{
+    id                            = "tfstate-expire"
+    enabled                       = true
+    expiration                    = 90
```
Comment (sblack4) [inline variables.tf]: Wait... does this just delete everything after 90 days?

**Review by sblack4 — CHANGES_REQUESTED** [review]

File: assumerole.tf (hunk 2 above)
Comment (cdaniluk) [reply assumerole.tf:21]: there's no chicken/egg problem here. This is for cross account scenarios only and is not intended to hold the state of the account holding this bucket. In single account scenarios and for the "parent" or "delegated" account that will hold tfstate, the CF Stack is the preferred option.

File: assumerole.tf (hunk 1 above)
Comment (cdaniluk) [reply assumerole.tf:8]: its a convenience script to run by hand. I thought it would save time and reduce mistakes.

File: variables.tf (hunk 3 above)
Comment (cdaniluk) [reply variables.tf]: yes, I thought we evicted this horrible, awful default a long time ago. I've nixed it.

**Review by sdickenson — APPROVED** [review]

File: assumerole.tf [hunk 4]
```diff
@@ -0,0 +1,36 @@
+locals {
+  external_id = coalesce(var.assumerole_role_external_id, random_password.external_id.result)
```
Comment (sblack4) [inline assumerole.tf:2]: Why? Why give a random password the ability to assume admin access to your parent account?

File: variables.tf [hunk 5]
```diff
@@ -1,52 +1,88 @@
...
+  description = "External ID to attach to role (this is required, a random ID will be generated if not specified here)"
   type        = string
 }
 
-variable "tags" {
-  default     = {}
-  description = "Mapping of any extra tags you want added to resources"
-  type        = map(string)
+variable "assumerole_role_attach_policies" {
+  default     = ["arn:aws:iam::aws:policy/AdministratorAccess"]
```
Comment (sblack4) [inline variables.tf:79]: This role is just to access the state, right? This seems excessive

**Review by sblack4 — APPROVED** [review]
I'd like to see:
- More explicit usage instructions
- An output for the CFN stack 
- Pare down default IAM
- lifecycle rules to retire old versions 
These can be future things

File: variables.tf (hunk 5 above)
Comment (cdaniluk) [reply variables.tf:79]: No, this is the assumed role in the sub account and Terraform effectively requires AdministratorAccess to do things. You can create a custom role that only allows services you expect to manage with Terraform, but that sounds like a hot mess to me.

File: assumerole.tf (hunk 4 above)
Comment (cdaniluk) [reply assumerole.tf:2]: The parent account de facto has such access anyway. This is just explicitly connecting the account intended for terraform usage to the subset of users that we decide should be able to run terraform in the subaccounts.

The external ID is a recommended best practice. It doesn't confer access to anything all by itself. It does the opposite, actually.. if you're granted access to assume this role in the subaccount, you also need to "know" the external ID to use it.

**Comment by cdaniluk** [comment]
> I'd like to see:
> 
>     * More explicit usage instructions
> 
>     * An output for the CFN stack
> 
>     * Pare down default IAM
> 
>     * lifecycle rules to retire old versions
>       These can be future things

I addressed some of these in your other comments but two questions..

explicit usage instructions - did you read https://learn.hashicorp.com/tutorials/terraform/aws-assumerole? I reference to that in the README. I view this as a reference implementation of Hashi's recommendation for cross account state management. So I didn't think it made sense to add a lot more detail than that. Happy to add some if you think its still relevant, though.

Output for CFN stack - are you thinking a TF output for it? Because that wouldn't make sense.. the stacks are created in the subaccounts by hand running the script. You have to do it that way to address the chicken/egg problem. Until the subaccount has a role assumable by the parent account, it is not terraform managed.

**Comment by sblack4** [comment]
> > I'd like to see:
> > ```
> > * More explicit usage instructions
> > 
> > * An output for the CFN stack
> > 
> > * Pare down default IAM
> > 
> > * lifecycle rules to retire old versions
> >   These can be future things
> > ```
> 
> I addressed some of these in your other comments but two questions..
> 
> explicit usage instructions - did you read https://learn.hashicorp.com/tutorials/terraform/aws-assumerole? I reference to that in the README. I view this as a reference implementation of Hashi's recommendation for cross account state management. So I didn't think it made sense to add a lot more detail than that. Happy to add some if you think its still relevant, though.
> 
> Output for CFN stack - are you thinking a TF output for it? Because that wouldn't make sense.. the stacks are created in the subaccounts by hand running the script. You have to do it that way to address the chicken/egg problem. Until the subaccount has a role assumable by the parent account, it is not terraform managed.

I was imagining you'd output that script and CFN template, log into the other account, then run it. I guess that doesn't make a ton of sense to make them an output when they're going to be sitting right in the same directory. The documentation I was talking about was something to document those steps. ie run this TF in the parent account, then run the script/CFN in the child account.

**Pull request merged**